app/Http/Controllers/DemandController.php

```diff
@@ -0,0 +1,334 @@
+        $log->ip = $logIp;
+        $log->event_type = "request";
+        $log->save();
+
+        $response->headers->set('X-Adshares-Cid', $cid);
+        $response->headers->set('X-Adshares-Lid', $log->id);
+
+        if (! $response->isNotModified($request)) {
+            $response->headers->set('Content-Type', ($isIECompat ? 'text/base64,' : '') . $mime);
+        }
+        return $response;
+    }
+
+    public function viewScript(Request $request)
+    {
+        $params = [json_encode($request->getSchemeAndHttpHost())];
+
+        $jsPath = config('app.env') == 'production' ? resource_path('assets/js/tmp-copy/view.x.js') : resource_path('assets/js/tmp-copy/view.min.js');
+
+        $response = new StreamedResponse();
+        $response->setCallback(function () use ($jsPath, $request, $params) {
+            echo str_replace([
+              "'{{ ORIGIN }}'",
+            ], $params, file_get_contents($jsPath));
+        });
+
+        $response->headers->set('Content-Type', 'text/javascript');
+
+        $response->setCache(array(
+          'etag' => md5(md5_file($jsPath) . implode(':', $params)),
+          'last_modified' => new \DateTime('@' . filemtime($jsPath)),
+          'max_age' => 3600 * 24 * 30,
+          's_maxage' => 3600 * 24 * 30,
+          'private' => false,
+          'public' => true
+        ));
+
+        if (! $response->isNotModified($request)) {
+            // TODO: ask Jacek
+        }
+        return $response;
+    }
+
+    /**
+     * @Route("/click/{id}", name="log_click", methods={"GET"}, requirements={"id": "\d+"})
+     *
+     */
+    public function clickAction(Request $request, $id)
+    {
+        $banner = Banner::getRepository($this->getDoctrine()->getManager())->find($id);
+        if (! $banner) {
+            throw new NotFoundHttpException();
+        }
+
+        $campaign = Campaign::getRepository($this->getDoctrine()->getManager())->find($banner->getCampaignId());
+        if (! $campaign) {
+            throw new NotFoundHttpException();
+        }
+
+        $url = $campaign->getLandingUrl();
+        $logIp = bin2hex(inet_pton($request->getClientIp()));
+
+        $cid = Utils::getRawTrackingId($request->query->get('cid'));
+        $pid = $request->query->get('pid');
+        $tid = Utils::getRawTrackingId($request->cookies->get('tid')) ?: $logIp;
+        $payTo = $request->query->get('pto');
+
+        $em = $this->getDoctrine()->getManager();
+        $log = new EventLog();
+        $log->cid = $cid;
+        $log->publisher_event_id = $pid;
+        $log->banner_id = $id;
+        $log->tid = $tid;
+        $log->ip = $logIp;
+        $log->pay_to = $payTo;
+        $log->event_type = "click";
+        $log->setTheirContext(Utils::getImpressionContext($this->container, $request));
+
+        $em->persist($log);
+        $em->flush();
+
+        $adpayService = $this->container->has('adpay') ? $this->container->get('adpay') : null;
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "adpay" 6 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)